### PR TITLE
8276660: Scalability bottleneck in java.security.Provider.getService()

### PR DESCRIPTION
**Pull request by lutkerd**
Backporting [JDK-8276660](https://bugs.openjdk.org/browse/JDK-8276660) which also fixes a bug in SecureRandom registration that was exposed during testing of https://github.com/corretto/amazon-corretto-crypto-provider. Currently, if you register a new provider then `new SecureRandom` doesn't work right away and only after making some other call like `Cipher.getInstance()` does it return the ACCP provided SecureRandom.

Ran all "java/security/Provider/", "jdk/jfr/event/" and tier1 tests locally on Linux x64 and they passed.

"[JDK-8254711](https://bugs.openjdk.org/browse/JDK-8254711): Add java.security.Provider.getService JFR Event" makes changes some of the same functioned but was different in 17u than upstream. To get a cleaner backport I reverted it locally, backported [JDK-8276660](https://bugs.openjdk.org/browse/JDK-8276660) and then re-applied 8254711 to get the updated code around the Map changes.

Also backporting the followup fixes:
8292739: Invalid legacy entries may be returned by Provider.getServic… 
8279222: Incorrect legacyMap.get in java.security.Provider after JDK-8276660

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8292739](https://bugs.openjdk.org/browse/JDK-8292739) needs maintainer approval
- [x] [JDK-8279222](https://bugs.openjdk.org/browse/JDK-8279222) needs maintainer approval
- [x] [JDK-8276660](https://bugs.openjdk.org/browse/JDK-8276660) needs maintainer approval

### Issues
 * [JDK-8276660](https://bugs.openjdk.org/browse/JDK-8276660): Scalability bottleneck in java.security.Provider.getService() (**Enhancement** - P3 - Approved)
 * [JDK-8292739](https://bugs.openjdk.org/browse/JDK-8292739): Invalid legacy entries may be returned by Provider.getServices() call (**Bug** - P4 - Approved)
 * [JDK-8279222](https://bugs.openjdk.org/browse/JDK-8279222): Incorrect legacyMap.get in java.security.Provider after JDK-8276660 (**Bug** - P4 - Approved)

### Reviewers
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/2730/head:pull/2730` \
`$ git checkout pull/2730`

Update a local copy of the PR: \
`$ git checkout pull/2730` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/2730/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2730`

View PR using the GUI difftool: \
`$ git pr show -t 2730`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/2730.diff">https://git.openjdk.org/jdk17u-dev/pull/2730.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/2730#issuecomment-2243373896)

### CODE REVIEW

File: src/java.base/share/classes/java/security/Provider.java [hunk 1]
```diff
@@ -1231,23 +1276,18 @@ private void parseLegacyPut(String name, String value) {
...
+        Service s = serviceMap.get(key);
         if (s == null) {
-            synchronized (this) {
-                ensureLegacyParsed();
-                if (legacyMap != null && !legacyMap.isEmpty()) {
-                    s = legacyMap.get(key);
-                }
+            s = legacyMap.get(key);
+            if (s != null && !s.isValid()) {
+                legacyMap.remove(key, s);
```
Comment (shipilev) [inline src/java.base/share/classes/java/security/Provider.java:1290]: This hunk seems to be different from the hunk in the original changeset: https://github.com/openjdk/jdk/commit/9b747491de01fd011b09668a67113e80c2b7c708#diff-c2847a8d1997d3d28a13354b05d5e21ac6656e7e17f849bc198715099b9b5bd2L1204-R1286 -- and the follow-ups did not change it as well. Does this stack of patches contain some other patch?

**Review by shipilev — COMMENTED** [review]
This looks okay, but I have questions:

File: src/java.base/share/classes/java/security/Provider.java (hunk 1 above)
Comment (lutkerd) [reply src/java.base/share/classes/java/security/Provider.java:1290]: This block comes from the JFR patch I reverted and re-applied: [https://github.com/openjdk/jdk/commit/bc2af47e1e4e6e96020e03a60ce098ddd17f63ba#diff-c2847a8d1997d3d28a13354b05d5[…]f849bc198715099b9b5bd2R1286-R1290](https://github.com/openjdk/jdk/commit/bc2af47e1e4e6e96020e03a60ce098ddd17f63ba#diff-c2847a8d1997d3d28a13354b05d5e21ac6656e7e17f849bc198715099b9b5bd2R1286-R1290)

**Comment by lutkerd** [comment]
@coffeys @valeriepeng, can you take a look?

File: src/java.base/share/classes/java/security/Provider.java (hunk 1 above)
Comment (shipilev) [reply src/java.base/share/classes/java/security/Provider.java:1290]: All right, thanks.

**Comment by coffeys** [comment]
> @coffeys @valeriepeng, can you take a look?

Changes look ok to me.

**Review by coffeys — APPROVED** [review]

**Comment by lutkerd** [comment]
/approval request Backporting JDK-8276660 to fix issue with Provider registration not taking immediate affect. Ran tier1-4 and reproducer of the issue without a problem. Also pulling in followups of JDK-8292739 and JDK-8279222 as to not introduce bugs. This was not clean and required reverting and re-applying JDK-8254711 to get the correct code.

**Pull request closed without merging**